File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/LibraryResourceLogic.kt
package com.lextures.android.core.lms

/** Access resolution for HE library / e-reserve resources (M3.6). */
sealed class LibraryAccessState {
    data class Ready(val url: String) : LibraryAccessState()
    data class Gated(val messageKey: String) : LibraryAccessState()
    data class RequiresWeb(val path: String) : LibraryAccessState()
}

enum class LibraryBrowseTab {
    Library,
    Oer,
}

object LibraryResourceLogic {
    fun libraryItems(items: List<CourseStructureItem>): List<CourseStructureItem> =
        items.filter { it.kind == "library_resource" }

    fun hasLibraryResources(items: List<CourseStructureItem>): Boolean =
        libraryItems(items).isNotEmpty()

    fun accessEventPath(courseCode: String, itemId: String): String =
        "/api/v1/courses/${encodePath(courseCode)}/library-resources/${encodePath(itemId)}/access"

    fun resourceDetailPath(courseCode: String, itemId: String): String =
        "/api/v1/courses/${encodePath(courseCode)}/library-resources/${encodePath(itemId)}"

    fun webModulePath(courseCode: String, itemId: String): String =
        "/courses/${encodePath(courseCode)}/modules/library-resource/${encodePath(itemId)}"

    fun resolveAccess(payload: LibraryResourcePayload): LibraryAccessState {
        normalizedUrl(payload.ezproxyUrl)?.let { return LibraryAccessState.Ready(it) }
        normalizedUrl(payload.metadata?.ezproxyUrl)?.let { return LibraryAccessState.Ready(it) }
        return when (payload.resourceType) {
            "leganto_list" -> LibraryAccessState.Gated("mobile.library.legantoGated")
            "catalog_item" -> LibraryAccessState.Gated("mobile.library.catalogGated")
            else -> LibraryAccessState.Gated("mobile.library.noAccess")
        }
    }

    fun resourceTypeLabel(resourceType: String): String = when (resourceType) {
        "leganto_list" -> "mobile.library.type.leganto"
        "catalog_item" -> "mobile.library.type.catalog"
        else -> "mobile.library.type.generic"
    }

    fun defaultOerProvider(providers: List<String>): String? {
        val preferred = listOf("oer_commons", "openstax", "merlot")
        return preferred.firstOrNull { it in providers } ?: providers.firstOrNull()
    }

    fun oerProviderLabel(provider: String): String = when (provider) {
        "oer_commons" -> "OER Commons"
        "openstax" -> "OpenStax"
        "merlot" -> "MERLOT"
        else -> provider
    }

    private fun normalizedUrl(raw: String?): String? {
        val trimmed = raw?.trim().orEmpty()
        return trimmed.ifEmpty { null }
    }

    private fun encodePath(value: String): String =
        java.net.URLEncoder.encode(value, "UTF-8").replace("+", "%20")
}

File: clients/ios/Lextures/Core/LMS/LibraryResourceLogic.swift
import Foundation

enum LibraryAccessState: Equatable {
    case ready(url: String)
    case gated(messageKey: String)
    case requiresWeb(path: String)
}

enum LibraryBrowseTab: String, CaseIterable, Identifiable {
    case library
    case oer

    var id: String { rawValue }
}

enum LibraryResourceLogic {
    static func libraryItems(from items: [CourseStructureItem]) -> [CourseStructureItem] {
        items.filter { $0.kind == "library_resource" }
    }

    static func hasLibraryResources(in items: [CourseStructureItem]) -> Bool {
        !libraryItems(from: items).isEmpty
    }

    static func accessEventPath(courseCode: String, itemId: String) -> String {
        "/api/v1/courses/\(LMSAPI.encodePath(courseCode))/library-resources/\(LMSAPI.encodePath(itemId))/access"
    }

    static func resourceDetailPath(courseCode: String, itemId: String) -> String {
        "/api/v1/courses/\(LMSAPI.encodePath(courseCode))/library-resources/\(LMSAPI.encodePath(itemId))"
    }

    static func webModulePath(courseCode: String, itemId: String) -> String {
        "/courses/\(LMSAPI.encodePath(courseCode))/modules/library-resource/\(LMSAPI.encodePath(itemId))"
    }

    static func resolveAccess(payload: LibraryResourcePayload) -> LibraryAccessState {
        if let url = normalizedURL(payload.ezproxyUrl) {
            return .ready(url: url)
        }
        if let url = normalizedURL(payload.metadata?.ezproxyUrl) {
            return .ready(url: url)
        }
        if payload.resourceType == "leganto_list" {
            return .gated(messageKey: "mobile.library.legantoGated")
        }
        if payload.resourceType == "catalog_item" {
            return .gated(messageKey: "mobile.library.catalogGated")
        }
        return .gated(messageKey: "mobile.library.noAccess")
    }

    static func resourceTypeLabel(_ resourceType: String) -> String {
        switch resourceType {
        case "leganto_list": return L.text("mobile.library.type.leganto")
        case "catalog_item": return L.text("mobile.library.type.catalog")
        default: return L.text("mobile.library.type.generic")
        }
    }

    static func subtitle(for item: CourseStructureItem, payload: LibraryResourcePayload?) -> String {
        if let payload {
            return resourceTypeLabel(payload.resourceType)
        }
        return L.text("mobile.library.type.generic")
    }

    static func parseLibrarySearchResults(from data: Data) throws -> [LibraryCatalogResult] {
        let response = try JSONDecoder().decode(LibrarySearchResponse.self, from: data)
        return response.results
    }

    static func parseOERSearchResults(from data: Data) throws -> OERSearchResponse {
        try JSONDecoder().decode(OERSearchResponse.self, from: data)
    }

    static func defaultOERProvider(from providers: [String]) -> String? {
        let preferred = ["oer_commons", "openstax", "merlot"]
        for id in preferred where providers.contains(id) {
            return id
        }
        return providers.first
    }

    static func oerProviderLabel(_ provider: String) -> String {
        switch provider {
        case "oer_commons": return "OER Commons"
        case "openstax": return "OpenStax"
        case "merlot": return "MERLOT"
        default: return provider
        }
    }

    private static func normalizedURL(_ raw: String?) -> String? {
        guard let raw else { return nil }
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}